glusterd: pack volume info status flags into a bitfield

dmantipov opened the pull request:
Pack georep status volume flag among other boolean flags
into the only bitfield within `glusterd_volinfo_t`, thus
dropping `gsync_status_param_t` quirk and simplifying
related georep code.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/xlators/mgmt/glusterd/src/glusterd-geo-rep.c b/xlators/mgmt/glusterd/src/glusterd-geo-rep.c
index 2034d6b7a..a8b011821 100644
--- a/xlators/mgmt/glusterd/src/glusterd-geo-rep.c
+++ b/xlators/mgmt/glusterd/src/glusterd-geo-rep.c
@@ -2147,8 +2147,8 @@ glusterd_check_geo_rep_running(glusterd_volinfo_t *volinfo, char **op_errstr)
     glusterd_check_geo_rep_configured(volinfo, &enabled);
 
     if (enabled) {
-        ret = dict_foreach(volinfo->gsync_secondaries,
-                           _get_secondary_status, volinfo);
+        ret = dict_foreach(volinfo->gsync_secondaries, _get_secondary_status,
+                           volinfo);
         if (ret) {
             gf_msg(THIS->name, GF_LOG_ERROR, 0,
                    GD_MSG_SECONDARYINFO_FETCH_ERROR,
diff --git a/xlators/mgmt/glusterd/src/glusterd-snapshot-utils.c b/xlators/mgmt/glusterd/src/glusterd-snapshot-utils.c
index f8ba8e324..e552fc31e 100644
--- a/xlators/mgmt/glusterd/src/glusterd-snapshot-utils.c
+++ b/xlators/mgmt/glusterd/src/glusterd-snapshot-utils.c
@@ -682,7 +682,8 @@ gd_add_vol_snap_details_to_dict(dict_t *dict, char *prefix,
     }
 
     snprintf(key, sizeof(key), "%s.is_snap_volume", prefix);
-    ret = dict_set_uint32(dict, key, !!(volinfo->flags & GD_VOLINFO_SNAP_VOLUME));
+    ret = dict_set_uint32(dict, key,
+                          !!(volinfo->flags & GD_VOLINFO_SNAP_VOLUME));
     if (ret) {
         gf_msg(this->name, GF_LOG_ERROR, 0, GD_MSG_DICT_SET_FAILED,
                "Unable to set %s for volume"
diff --git a/xlators/mgmt/glusterd/src/glusterd-volgen.c b/xlators/mgmt/glusterd/src/glusterd-volgen.c
index 21b0a4b7e..336deabba 100644
--- a/xlators/mgmt/glusterd/src/glusterd-volgen.c
+++ b/xlators/mgmt/glusterd/src/glusterd-volgen.c
@@ -2497,7 +2497,8 @@ brick_graph_add_server(volgen_graph_t *graph, glusterd_volinfo_t *volinfo,
         return -1;
 
     volname = (volinfo->flags & GD_VOLINFO_SNAP_VOLUME)
-        ? volinfo->parent_volname : volinfo->volname;
+                  ? volinfo->parent_volname
+                  : volinfo->volname;
 
     if (volname && !strcmp(volname, GLUSTER_SHARED_STORAGE)) {
         ret = xlator_set_fixed_option(xl, "strict-auth-accept", "true");
@@ -5776,7 +5777,8 @@ generate_client_volfiles(glusterd_volinfo_t *volinfo,
     gf_transport_type type = GF_TRANSPORT_TCP;
 
     volname = (volinfo->flags & GD_VOLINFO_SNAP_VOLUME)
-        ? volinfo->parent_volname : volinfo->volname;
+                  ? volinfo->parent_volname
+                  : volinfo->volname;
 
     if (volname && !strcmp(volname, GLUSTER_SHARED_STORAGE) &&
         client_type != GF_CLIENT_TRUSTED) {
```

dmantipov commented:
/run regression

mykaul commented:
I personally like this direction, the community was less enthusiastic - see the discussion @ https://github.com/gluster/glusterfs/pull/2929

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-op-sm.c:
```diff
@@ -2948,12 +2948,14 @@ glusterd_op_set_volume(dict_t *dict, char **errstr)
 
         if (strcmp(key, "config.memory-accounting") == 0) {
-            ret = gf_string2boolean(value, &volinfo->memory_accounting);
+            gf_boolean_t val = _gf_false;
```
I think `GF_VOLINFO_MEMORY_ACCOUNTING` could be already set here, so we need to clear it when the boolean is false.

dmantipov replied:
OK

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-snapshot-utils.c:
```diff
@@ -1075,7 +1076,8 @@ gd_import_volume_snap_details(dict_t *dict, glusterd_volinfo_t *volinfo,
         goto out;
     }
-    volinfo->is_snap_volume = (is_snap_int != 0);
+    if (is_snap_int != 0)
```
Are we sure `GD_VOLINFO_SNAP_VOLUME` will never be set here ? otherwise we need to clear it if this is false.

dmantipov replied:
Hopefully OK as well.

xhernandez reviewed (commented):
I'm ok with the approach. However two things:

- Probably it would be good to write small inline wrappers to check flags. This can make the code cleaner and more readable.

- Glusterd has had several issues in the past because it's not thread-safe in some cases. Updating independent variables is always safe, but updating independent bits in the same variable may be problematic. I don't know if there can be concurrent updates in some case, though. @mohit84 @nik-redhat do you know if this is safe ?

@amarts @pranithk are you also ok with this approach ?

Commit:
glusterd: pack volume info status flags into a bitfield

Pack georep status volume flag among other boolean flags
into the only bitfield within 'glusterd_volinfo_t', thus
dropping 'gsync_status_param_t' quirk and simplifying
related georep code.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

dmantipov commented:
Hopefully `GF_DECLARE_BITOP()` is generic enough to be used everywhere.

dmantipov commented:
/run regression

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/common-utils.h b/libglusterfs/src/glusterfs/common-utils.h
index 9b57bc3ae..f6ed619d1 100644
--- a/libglusterfs/src/glusterfs/common-utils.h
+++ b/libglusterfs/src/glusterfs/common-utils.h
@@ -209,7 +209,9 @@ gf_set_nofile(rlim_t high, rlim_t low);
 
 #else /* Darwin */
 
-#define gf_set_nofile(low, high) do { } while (0)
+#define gf_set_nofile(low, high)                                               \
+    do {                                                                       \
+    } while (0)
 
 #endif /* not GF_DARWIN_HOST_OS */
 
@@ -478,19 +480,16 @@ union gf_sock_union {
 
 /* Macro to declare basic operations on a bitfield. */
 
-#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                     \
-                                                                              \
-static inline gf_boolean_t ptr##_has_##suffix(type *ptr) {                    \
-    return ptr->member & flag;                                                \
-}                                                                             \
-                                                                              \
-static inline void ptr##_set_##suffix(type *ptr) {                            \
-    ptr->member |= flag;                                                      \
-}                                                                             \
-                                                                              \
-static inline void ptr##_clear_##suffix(type *ptr) {                          \
-    ptr->member &= ~flag;                                                     \
-}                                                                             \
+#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                      \
+                                                                               \
+    static inline gf_boolean_t ptr##_has_##suffix(type *ptr)                   \
+    {                                                                          \
+        return ptr->member & flag;                                             \
+    }                                                                          \
+                                                                               \
+    static inline void ptr##_set_##suffix(type *ptr) { ptr->member |= flag; }  \
+                                                                               \
+    static inline void ptr##_clear_##suffix(type *ptr) { ptr->member &= ~flag; }
 
 static inline gf_boolean_t
 gf_irrelevant_entry(struct dirent *entry)
```

dmantipov commented:
/recheck smoke

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/common-utils.h b/libglusterfs/src/glusterfs/common-utils.h
index 9b57bc3ae..f6ed619d1 100644
--- a/libglusterfs/src/glusterfs/common-utils.h
+++ b/libglusterfs/src/glusterfs/common-utils.h
@@ -209,7 +209,9 @@ gf_set_nofile(rlim_t high, rlim_t low);
 
 #else /* Darwin */
 
-#define gf_set_nofile(low, high) do { } while (0)
+#define gf_set_nofile(low, high)                                               \
+    do {                                                                       \
+    } while (0)
 
 #endif /* not GF_DARWIN_HOST_OS */
 
@@ -478,19 +480,16 @@ union gf_sock_union {
 
 /* Macro to declare basic operations on a bitfield. */
 
-#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                     \
-                                                                              \
-static inline gf_boolean_t ptr##_has_##suffix(type *ptr) {                    \
-    return ptr->member & flag;                                                \
-}                                                                             \
-                                                                              \
-static inline void ptr##_set_##suffix(type *ptr) {                            \
-    ptr->member |= flag;                                                      \
-}                                                                             \
-                                                                              \
-static inline void ptr##_clear_##suffix(type *ptr) {                          \
-    ptr->member &= ~flag;                                                     \
-}                                                                             \
+#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                      \
+                                                                               \
+    static inline gf_boolean_t ptr##_has_##suffix(type *ptr)                   \
+    {                                                                          \
+        return ptr->member & flag;                                             \
+    }                                                                          \
+                                                                               \
+    static inline void ptr##_set_##suffix(type *ptr) { ptr->member |= flag; }  \
+                                                                               \
+    static inline void ptr##_clear_##suffix(type *ptr) { ptr->member &= ~flag; }
 
 static inline gf_boolean_t
 gf_irrelevant_entry(struct dirent *entry)
```

dmantipov commented:
/recheck smoke

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/common-utils.h b/libglusterfs/src/glusterfs/common-utils.h
index 9b57bc3ae..f6ed619d1 100644
--- a/libglusterfs/src/glusterfs/common-utils.h
+++ b/libglusterfs/src/glusterfs/common-utils.h
@@ -209,7 +209,9 @@ gf_set_nofile(rlim_t high, rlim_t low);
 
 #else /* Darwin */
 
-#define gf_set_nofile(low, high) do { } while (0)
+#define gf_set_nofile(low, high)                                               \
+    do {                                                                       \
+    } while (0)
 
 #endif /* not GF_DARWIN_HOST_OS */
 
@@ -478,19 +480,16 @@ union gf_sock_union {
 
 /* Macro to declare basic operations on a bitfield. */
 
-#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                     \
-                                                                              \
-static inline gf_boolean_t ptr##_has_##suffix(type *ptr) {                    \
-    return ptr->member & flag;                                                \
-}                                                                             \
-                                                                              \
-static inline void ptr##_set_##suffix(type *ptr) {                            \
-    ptr->member |= flag;                                                      \
-}                                                                             \
-                                                                              \
-static inline void ptr##_clear_##suffix(type *ptr) {                          \
-    ptr->member &= ~flag;                                                     \
-}                                                                             \
+#define GF_DECLARE_BITOP(type, ptr, member, flag, suffix)                      \
+                                                                               \
+    static inline gf_boolean_t ptr##_has_##suffix(type *ptr)                   \
+    {                                                                          \
+        return ptr->member & flag;                                             \
+    }                                                                          \
+                                                                               \
+    static inline void ptr##_set_##suffix(type *ptr) { ptr->member |= flag; }  \
+                                                                               \
+    static inline void ptr##_clear_##suffix(type *ptr) { ptr->member &= ~flag; }
 
 static inline gf_boolean_t
 gf_irrelevant_entry(struct dirent *entry)
```

dmantipov commented:
/run regression
